### PR TITLE
Reduce memory usage in crossref.rs by sharing strings

### PR DESCRIPTION
**Pull request by amccreight**
(This includes my patch for bug 1416876 because it is based on top of that and I don't know how to do that the right way.)

I did some light local testing, all the way up to running a webserver for the results of a full Firefox run and the few things I checked seemed ok. The output from this file was also unchanged, when I checked it for the small set of test files.

### CODE REVIEW

File: tools/src/bin/crossref.rs [hunk 1]
```diff
@@ -61,6 +64,28 @@ fn split_scopes(id: &str) -> Vec<String> {
     return result;
 }
 
+struct StringIntern {
+    set: HashMap<Rc<String>, ()>
```
Comment (emilio) [inline tools/src/bin/crossref.rs]: nit: Isn't this basically a `HashSet<Rc<String>>`?

**Comment by emilio** [comment]
(We could also consider using https://github.com/servo/string-cache, instead of a hand-rolled one)

File: tools/src/bin/crossref.rs (hunk 1 above)
Comment (amccreight) [reply tools/src/bin/crossref.rs]: As far as I can tell, HashSet<> doesn't support the Entry interface, or anything like that. I want to use the existing value if there is one, insert our new value if there isn't, and avoid repeated lookups. I guess that's overoptimizing. I could use a HashSet and just do two lookups.

**Comment by amccreight** [comment]
I've looked at the Servo string cache before, but I think it is threadsafe and it was slower than the trivial thing when I tried it before. I also think just writing a tiny 6 line function is simpler than introducing a new dependency with its own API people would have to understand.

File: tools/src/bin/crossref.rs (hunk 1 above)
Comment (emilio) [reply tools/src/bin/crossref.rs]: Well, that's fair... Sounds ok :)

**Comment by emilio** [comment]
> I've looked at the Servo string cache before, but I think it is threadsafe and it was slower than the trivial thing when I tried it before. I also think just writing a tiny 6 line function is simpler than introducing a new dependency with its own API people would have to understand.

I mean, if we're really going to uber-optimize we can also stop using siphash and all that stuff, but yeah, this is fine, sorry for the contextless review, and thanks for saving my laptop's memory ;)

**Comment by amccreight** [comment]
I actually did write a patch to replace the hash function in the string interner, but there was no impact on performance. I vaguely recall reading that the default hash function is only slow on small keys, and we're using strings, so maybe it isn't an issue? Anyways, I should just profile this thing to see if there are some easy wins in terms of speed.

**Pull request merged**